File: ValidParentheses.java
import java.util.*;

class Solution {
    boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '{' || c == '[' || c == '(') {
                stack.push(c);
            } else {
                if (stack.isEmpty())
                    return false;
                char temp = stack.pop();
                if (c == '}' && temp != '{' || c == ']' && temp != '[' || c == ')' && temp != '(')
                    return false;
            }
        }
        return stack.isEmpty();
    }
    /*
     * *
     * bool isValid(string s) {
     * char[] stack = new char[s.length()];
     * int t=-1;
     * for(char c: s.toCharArray()){
     * if (c == '{' || c == '[' || c == '(') {
     * stack[++t] = c;
     * } else {
     * char temp = stack[t--];
     * if (c == '}' && temp != '{' || c == '}' && temp != '{' || c == '}' && temp !=
     * '{') {
     * return false;
     * }
     * }
     * }
     * return t == -1;
     * }
     * }
     */
}

File: ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c == "{" or c == "[" or c == "(":
                stack.append(c)
            else:
                if len(stack) == 0:
                    return False
                temp = stack.pop()
                if (
                    c == "}"
                    and temp != "{"
                    or c == "]"
                    and temp != "["
                    or c == ")"
                    and temp != "("
                ):
                    return False
        return len(stack) == 0
